addons/stock_picking_batch/models/stock_picking.py

```diff
@@ -82,7 +112,100 @@ def action_add_operations(self):
+
+        # If nothing was found after those two steps, then no batch is doable given the conditions
+        return False
+
+    def _is_auto_batchable(self, picking=None):
+        """ Verifies if a picking can be put in a batch with another picking without violating auto_batch constrains.
+        """
+        res = True
+        if not picking:
+            picking = self.env['stock.picking']
+        if self.picking_type_id.batch_max_lines:
+            res = res and (len(self.move_ids) + len(picking.move_ids) <= self.picking_type_id.batch_max_lines)
+        if self.picking_type_id.batch_max_pickings:
+            # Sounds absurd. BUT if we put "batch max picking" to a value <= 1, makes sense ... Or not. Because then there is no point to batch.
+            res = res and self.picking_type_id.batch_max_pickings > 1
+        return res
+
+    def _get_possible_pickings_domain(self):
+        self.ensure_one()
+        domain = [
+            ('id', '!=', self.id),
+            ('company_id', '=', self.company_id.id if self.company_id else False),
+            ('immediate_transfer', '=', False),
+            ('state', 'in', ('waiting', 'confirmed', 'assigned')),
+            ('picking_type_id', '=', self.picking_type_id.id),
+            ('batch_id', '=', False),
+        ]
+        if self.picking_type_id.batch_group_by_partner:
+            domain = expression.AND([domain, [('partner_id', '=', self.partner_id.id)]])
+        if self.picking_type_id.batch_group_by_destination:
+            domain = expression.AND([domain, [('partner_id.country_id', '=', self.partner_id.country_id.id)]])
+        if self.picking_type_id.batch_group_by_src_loc:
+            domain = expression.AND([domain, [('location_id', '=', self.location_id.id)]])
+        if self.picking_type_id.batch_group_by_dest_loc:
+            domain = expression.AND([domain, [('location_dest_id', '=', self.location_dest_id.id)]])
+
+        return domain
+
+    def _get_possible_batches_domain(self):
+        self.ensure_one()
+        domain = [
+            ('state', 'in', ('draft', 'in_progress') if self.picking_type_id.batch_auto_confirm else ('draft',)),
+            ('picking_type_id', '=', self.picking_type_id.id),
+            ('company_id', '=', self.company_id.id if self.company_id else False),
+        ]
+        if self.picking_type_id.batch_group_by_partner:
+            domain = expression.AND([domain, [('picking_ids.partner_id', '=', self.partner_id.id)]])
+        if self.picking_type_id.batch_group_by_destination:
+            domain = expression.AND([domain, [('picking_ids.partner_id.country_id', '=', self.partner_id.country_id.id)]])
+        if self.picking_type_id.batch_group_by_src_loc:
+            domain = expression.AND([domain, [('picking_ids.location_id', '=', self.location_id.id)]])
+        if self.picking_type_id.batch_group_by_dest_loc:
+            domain = expression.AND([domain, [('picking_ids.location_dest_id', '=', self.location_dest_id.id)]])
```

Comment:
Since we will do that search on each picking it would be great to adapt also the index on batch picking